nuget/helpers/lib/NuGetUpdater/DotNetPackageCorrelation/Correlator.cs

```diff
@@ -0,0 +1,197 @@
+        foreach (var directory in Directory.EnumerateDirectories(_releaseNotesDirectory.FullName))
+        {
+            var directoryName = Path.GetFileName(directory);
+            if (Version.TryParse(directoryName, out var version))
+            {
+                runtimeVersions.Add(version);
+            }
+        }
+
+        var runtimePackages = new RuntimePackages();
+        var warnings = new List<string>();
+        foreach (var majorVersion in runtimeVersions)
+        {
+            var releasesJsonPath = Path.Combine(_releaseNotesDirectory.FullName, majorVersion.ToString(), "releases.json");
+            if (!File.Exists(releasesJsonPath))
+            {
+                warnings.Add($"Unable to find releases.json file for version {majorVersion}");
+                continue;
+            }
+
+            var releasesJson = await File.ReadAllTextAsync(releasesJsonPath);
+            var releasesFile = JsonSerializer.Deserialize<ReleasesFile>(releasesJson, SerializerOptions)!;
+
+            foreach (var release in releasesFile.Releases)
+            {
+                foreach (var sdk in release.GetSdks())
+                {
+                    if (sdk.Version is null)
+                    {
+                        warnings.Add($"Skipping release with missing version information from {releasesJson}");
+                        continue;
+                    }
+
+                    if (sdk.RuntimeVersion is null)
+                    {
+                        warnings.Add($"Skipping release with missing runtime version information from {releasesJson}");
+                        continue;
+                    }
+
+                    if (runtimePackages.Runtimes.ContainsKey(sdk.RuntimeVersion))
+                    {
+                        // already processed this runtime
+                        continue;
+                    }
+
+                    var packagesAndVersions = new PackageSet();
+                    runtimePackages.Runtimes.Add(sdk.RuntimeVersion, packagesAndVersions);
+                    var runtimeDirectory = new DirectoryInfo(Path.Combine(_releaseNotesDirectory.FullName, majorVersion.ToString(), sdk.RuntimeVersion.ToString()));
+                    var runtimeMarkdownPath = Path.Combine(runtimeDirectory.FullName, $"{sdk.RuntimeVersion}.md");
+                    if (!File.Exists(runtimeMarkdownPath))
+                    {
+                        warnings.Add($"Unable to find expected markdown file {runtimeMarkdownPath}");
+                        continue;
+                    }
+
+                    var markdownContent = await File.ReadAllTextAsync(runtimeMarkdownPath);
+                    var packages = GetPackagesFromMarkdown(runtimeMarkdownPath, markdownContent, warnings);
+                    foreach (var (packageName, packageVersion) in packages)
+                    {
+                        packagesAndVersions.Packages[packageName] = packageVersion;
```

Comment:
Is it possible to have multiple package versions within this set and is taking the "last" one the appropriate behavior if so?

Reply:
It's not possible to have multiple versions of the same package, but it is possible that we see a certain package twice if it had assemblies and analyzers, etc. so the overwrites are fine.